File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/user/MICLoginTest.java
package com.kinvey.androidTest.store.user;


import android.content.Context;
import android.os.Looper;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.SmallTest;
import androidx.test.platform.app.InstrumentationRegistry;

import com.google.api.client.http.UrlEncodedContent;
import com.kinvey.android.Client;
import com.kinvey.android.callback.KinveyMICCallback;
import com.kinvey.android.model.User;
import com.kinvey.android.store.UserStore;
import com.kinvey.androidTest.model.TestUser;
import com.kinvey.java.AbstractClient;
import com.kinvey.java.auth.Credential;
import com.kinvey.java.auth.KinveyAuthRequest;
import com.kinvey.java.store.UserStoreRequestManager;
import com.kinvey.java.store.requests.user.GetMICAccessToken;
import com.kinvey.java.store.requests.user.GetMICTempURL;
import com.kinvey.java.store.requests.user.LoginToTempURL;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.util.HashMap;
import java.util.concurrent.CountDownLatch;

import static com.kinvey.androidTest.TestManager.PASSWORD;
import static com.kinvey.androidTest.TestManager.USERNAME;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class MICLoginTest {

    private Client client = null;
    private static final String APP_KEY = "YOUR_APP_KEY_HERE";
    private static final String APP_SECRET = "YOUR_APP_SECRET_HERE";
    private static final String CLIENT_ID = "CLIENT_ID";
    private static final String REFRESH_TOKEN = "refresh_token";
    private static final String AUTH_TOKEN = "auth_token";
    private static final String REDIRECT_URI = "http://test.redirect";
    private static final String CLIENT_ID_FIELD = "client_id";


    @Before
    public void setup() throws InterruptedException {
        Context mMockContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        client = new Client.Builder(APP_KEY, APP_SECRET, mMockContext).setUserClass(TestUser.class).build();
    }

    // Check clientId in auth link for MICLoginPage
    @Test
    public void testMIC_LoginWithAuthorizationCodeLoginPageWithClientId() throws InterruptedException {
        DefaultKinveyMICCallback userCallback = loginWithAuthorizationCodeLoginPage(CLIENT_ID, REDIRECT_URI, client);
        assertNotNull(userCallback.myURLToRender);
        assertTrue(!userCallback.myURLToRender.isEmpty());
        assertTrue(userCallback.myURLToRender.startsWith(client.getMICHostName() + client.getMICApiVersion() + "/oauth/auth?client_id=" + APP_KEY + "." + CLIENT_ID));
    }

    // Check clientId (should be absent second part of client_id) in auth link for MICLoginPage
    @Test
    public void testMIC_LoginWithAuthorizationCodeLoginPage() throws InterruptedException {
        DefaultKinveyMICCallback userCallback = loginWithAuthorizationCodeLoginPage(null, REDIRECT_URI, client);
        assertNotNull(userCallback.myURLToRender);
        assertTrue(!userCallback.myURLToRender.isEmpty());
        assertTrue(userCallback.myURLToRender.startsWith(client.getMICHostName() + client.getMICApiVersion() + "/oauth/auth?client_id=" + APP_KEY + "&"));
    }

    // Check that myURLToRender contains openId parameter
    @Test
    public void testOpenIdExists() throws InterruptedException {
        DefaultKinveyMICCallback userCallback = loginWithAuthorizationCodeLoginPage(null, REDIRECT_URI, client);
        assertNotNull(userCallback.myURLToRender);
        assertTrue(!userCallback.myURLToRender.isEmpty());
        assertTrue(userCallback.myURLToRender.endsWith("&response_type=code" + "&scope=openid"));
    }

    private DefaultKinveyMICCallback loginWithAuthorizationCodeLoginPage(final String clientId, final String redirectUrl, final Client client) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyMICCallback callback = new DefaultKinveyMICCallback(latch);
        new Thread(new Runnable() {
            public void run() {
                Looper.prepare();
                UserStore.loginWithAuthorizationCodeLoginPage(client, clientId, redirectUrl, callback);
                Looper.loop();
            }
        }).start();
        latch.await();
        return callback;
    }

    // Check clientId for getting Temp Link
    @Test
    public void testGetMICTempURLWithClientId() throws IOException {
        UserStoreRequestManager requestManager = new UserStoreRequestManager(client, createBuilder(client));
        requestManager.setMICRedirectURI(REDIRECT_URI);
        GetMICTempURL micTempURL = requestManager.getMICTempURL(CLIENT_ID);
        assertTrue(micTempURL.getUriTemplate().endsWith("scope=openid"));
        assertEquals(APP_KEY + "." + CLIENT_ID, ((HashMap) ((UrlEncodedContent) micTempURL.getHttpContent()).getData()).get(CLIENT_ID_FIELD));
    }

    @Test
    public void testLoginToTempURL() throws IOException {
        UserStoreRequestManager requestManager = new UserStoreRequestManager(client, createBuilder(client));
        LoginToTempURL loginToTempURL = requestManager.MICLoginToTempURL(USERNAME, PASSWORD, CLIENT_ID, "tempURL");
        requestManager.setMICRedirectURI(REDIRECT_URI);
        GetMICTempURL micTempURL = requestManager.getMICTempURL(CLIENT_ID);
        assertTrue(micTempURL.getUriTemplate().endsWith("scope=openid"));
        assertEquals(APP_KEY + "." + CLIENT_ID, ((HashMap) ((UrlEncodedContent) loginToTempURL.getHttpContent()).getData()).get(CLIENT_ID_FIELD));
    }

    // Check clientId (should be absent second part of client_id) for getting Temp Link
    @Test
    public void testGetMICTempURL() throws IOException {
        UserStoreRequestManager requestManager = new UserStoreRequestManager(client, createBuilder(client));
        requestManager.setMICRedirectURI(REDIRECT_URI);
        GetMICTempURL micTempURL = requestManager.getMICTempURL(null);
        assertEquals(APP_KEY, ((HashMap) ((UrlEncodedContent) micTempURL.getHttpContent()).getData()).get(CLIENT_ID_FIELD));
    }

    // Check clientId for using refresh token
    @Test
    public void testMICLoginUseRefreshTokenWithClientId() throws IOException {
        UserStoreRequestManager requestManager = new UserStoreRequestManager(client, createBuilder(client));
        User user = new User();
        user.setId("userId");
        client.setActiveUser(user);
        Credential credential = new Credential(client.getActiveUser().getId(), AUTH_TOKEN, REFRESH_TOKEN);
        credential.setClientId(CLIENT_ID);
        client.getStore().store(client.getActiveUser().getId(), credential);
        GetMICAccessToken getMICAccessToken = requestManager.useRefreshToken(REFRESH_TOKEN);
        assertEquals(APP_KEY + "." + CLIENT_ID, ((HashMap) ((UrlEncodedContent) getMICAccessToken.getHttpContent()).getData()).get(CLIENT_ID_FIELD));
    }

    // Check clientId (should be absent second part of client_id) for using refresh token
    @Test
    public void testMICLoginUseRefreshToken() throws IOException {
        UserStoreRequestManager requestManager = new UserStoreRequestManager(client, createBuilder(client));
        User user = new User();
        user.setId("userId");
        client.setActiveUser(user);
        Credential credential = new Credential(client.getActiveUser().getId(), AUTH_TOKEN, REFRESH_TOKEN);
        client.getStore().store(client.getActiveUser().getId(), credential);
        GetMICAccessToken getMICAccessToken = requestManager.useRefreshToken(REFRESH_TOKEN);
        assertEquals(APP_KEY, ((HashMap) ((UrlEncodedContent) getMICAccessToken.getHttpContent()).getData()).get(CLIENT_ID_FIELD));
    }

    private KinveyAuthRequest.Builder createBuilder(AbstractClient client) {
        return new KinveyAuthRequest.Builder(client.getRequestFactory().getTransport(),
                client.getJsonFactory(), client.getBaseUrl(), APP_KEY, APP_SECRET, null);
    }


    private static class DefaultKinveyMICCallback implements KinveyMICCallback<User> {

        private CountDownLatch latch;
        private User result;
        private Throwable error;
        private String myURLToRender;

        private DefaultKinveyMICCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(User result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        private void finish() {
            latch.countDown();
        }

        @Override
        public void onReadyToRender(String myURLToRender) {
            this.myURLToRender = myURLToRender;
            finish();
        }
    }

    @Test
    public void testMICDefaultAPIVersion() throws IOException {
        assertEquals("v3", client.getMICApiVersion());
    }

    @Test
    public void testMICCustomAPIVersion() throws IOException {
        client.setMICApiVersion("1");
        assertEquals("v1", client.getMICApiVersion());
    }

}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/user/MICLoginTest.kt
package com.kinvey.androidTest.store.user

import android.os.Looper
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.google.api.client.http.UrlEncodedContent
import com.google.api.client.json.JsonFactory
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.callback.KinveyMICCallback
import com.kinvey.android.model.User
import com.kinvey.android.store.UserStore.Companion.loginWithAuthorizationCodeLoginPage
import com.kinvey.androidTest.TestManager.PASSWORD
import com.kinvey.androidTest.TestManager.USERNAME
import com.kinvey.androidTest.model.TestUser
import com.kinvey.java.AbstractClient
import com.kinvey.java.auth.Credential
import com.kinvey.java.auth.KinveyAuthRequest
import com.kinvey.java.store.UserStoreRequestManager
import junit.framework.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.util.*
import java.util.concurrent.CountDownLatch

@RunWith(AndroidJUnit4::class)
@SmallTest
class MICLoginTest {

    private var client: Client<TestUser>? = null

    @Before
    @Throws(InterruptedException::class)
    fun setup() {
        val mMockContext = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<TestUser>(APP_KEY, APP_SECRET, mMockContext).setUserClass(TestUser::class.java).build()
    }

    // Check clientId in auth link for MICLoginPage
    @Test
    @Throws(InterruptedException::class)
    fun testMIC_LoginWithAuthorizationCodeLoginPageWithClientId() {
        val userCallback = loginWithAuthorizationCodeLoginPage(CLIENT_ID, REDIRECT_URI, client)
        assertNotNull(userCallback.myURLToRender)
        assertTrue(!userCallback.myURLToRender.isNullOrEmpty())
        assertTrue(userCallback.myURLToRender?.startsWith(client?.micHostName + client?.micApiVersion + "/oauth/auth?client_id=" + APP_KEY + "." + CLIENT_ID) == true)
    }

    // Check clientId (should be absent second part of client_id) in auth link for MICLoginPage
    @Test
    @Throws(InterruptedException::class)
    fun testMIC_LoginWithAuthorizationCodeLoginPage() {
        val userCallback = loginWithAuthorizationCodeLoginPage(null, REDIRECT_URI, client)
        assertNotNull(userCallback.myURLToRender)
        assertTrue(!userCallback.myURLToRender.isNullOrEmpty())
        assertTrue(userCallback.myURLToRender?.startsWith(client?.micHostName + client?.micApiVersion + "/oauth/auth?client_id=" + APP_KEY + "&") == true)
    }

    // Check that myURLToRender contains openId parameter
    @Test
    @Throws(InterruptedException::class)
    fun testOpenIdExists() {
        val userCallback = loginWithAuthorizationCodeLoginPage(null, REDIRECT_URI, client)
        assertNotNull(userCallback.myURLToRender)
        assertTrue(!userCallback.myURLToRender.isNullOrEmpty())
        assertTrue(userCallback.myURLToRender?.endsWith("&response_type=code" + "&scope=openid") == true)
    }

    @Throws(InterruptedException::class)
    private fun loginWithAuthorizationCodeLoginPage(clientId: String?, redirectUrl: String, client: Client<*>?): DefaultKinveyMICCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyMICCallback(latch)
        Thread(Runnable {
            Looper.prepare()
            loginWithAuthorizationCodeLoginPage(client!!, clientId, redirectUrl, callback)
            Looper.loop()
        }).start()
        latch.await()
        return callback
    }

    // Check clientId for getting Temp Link
    @Test
    @Throws(IOException::class)
    fun testGetMICTempURLWithClientId() {
        val requestManager = UserStoreRequestManager(client, createBuilder(client) as KinveyAuthRequest.Builder<TestUser>)
        requestManager.micRedirectURI = REDIRECT_URI
        val micTempURL = requestManager.getMICTempURL(CLIENT_ID)
        assertTrue(micTempURL.uriTemplate?.endsWith("scope=openid") == true)
        assertEquals("$APP_KEY.$CLIENT_ID", ((micTempURL.httpContent as UrlEncodedContent?)?.data as HashMap<*, *>)[CLIENT_ID_FIELD])
    }

    @Test
    @Throws(IOException::class)
    fun testLoginToTempURL() {
        val requestManager = UserStoreRequestManager<TestUser>(client, createBuilder(client) as KinveyAuthRequest.Builder<TestUser>)
        val loginToTempURL = requestManager.MICLoginToTempURL(USERNAME, PASSWORD, CLIENT_ID, "tempURL")
        requestManager.micRedirectURI = REDIRECT_URI
        val micTempURL = requestManager.getMICTempURL(CLIENT_ID)
        assertTrue(micTempURL.uriTemplate?.endsWith("scope=openid") == true)
        assertEquals("$APP_KEY.$CLIENT_ID", ((loginToTempURL.httpContent as UrlEncodedContent?)?.data as HashMap<*, *>)[CLIENT_ID_FIELD])
    }

    // Check clientId (should be absent second part of client_id) for getting Temp Link
    @Test
    @Throws(IOException::class)
    fun testGetMICTempURL() {
        val requestManager = UserStoreRequestManager(client, createBuilder(client) as KinveyAuthRequest.Builder<TestUser>)
        requestManager.micRedirectURI = REDIRECT_URI
        val micTempURL = requestManager.getMICTempURL(null)
        assertEquals(APP_KEY, ((micTempURL.httpContent as UrlEncodedContent?)?.data as HashMap<*, *>)[CLIENT_ID_FIELD])
    }

    // Check clientId for using refresh token
    @Test
    @Throws(IOException::class)
    fun testMICLoginUseRefreshTokenWithClientId() {
        val requestManager = UserStoreRequestManager(client, createBuilder(client) as KinveyAuthRequest.Builder<TestUser>)
        val user = TestUser()
        user.id = "userId"
        client?.activeUser = user
        val credential = Credential(client?.activeUser?.id, AUTH_TOKEN, REFRESH_TOKEN)
        credential.clientId = CLIENT_ID
        client?.store?.store(client?.activeUser?.id, credential)
        val getMICAccessToken = requestManager.useRefreshToken(REFRESH_TOKEN)
        assertEquals("$APP_KEY.$CLIENT_ID", ((getMICAccessToken.httpContent as UrlEncodedContent?)?.data as HashMap<*, *>)[CLIENT_ID_FIELD])
    }

    // Check clientId (should be absent second part of client_id) for using refresh token
    @Test
    @Throws(IOException::class)
    fun testMICLoginUseRefreshToken() {
        val requestManager = UserStoreRequestManager(client, createBuilder(client) as KinveyAuthRequest.Builder<TestUser>)
        val user = TestUser()
        user.id = "userId"
        client?.activeUser = user
        val credential = Credential(client?.activeUser?.id, AUTH_TOKEN, REFRESH_TOKEN)
        client?.store?.store(client?.activeUser?.id, credential)
        val getMICAccessToken = requestManager.useRefreshToken(REFRESH_TOKEN)
        assertEquals(APP_KEY, ((getMICAccessToken.httpContent as UrlEncodedContent?)?.data as HashMap<*, *>)[CLIENT_ID_FIELD])
    }

    private fun createBuilder(client: AbstractClient<*>?): KinveyAuthRequest.Builder<*> {
        return KinveyAuthRequest.Builder<TestUser>(client?.requestFactory?.transport,
                client?.jsonFactory as JsonFactory, client?.baseUrl, APP_KEY, APP_SECRET, null)
    }

    private class DefaultKinveyMICCallback(private val latch: CountDownLatch) : KinveyMICCallback<User> {
        private var result: User? = null
        private var error: Throwable? = null
        var myURLToRender: String? = null

        override fun onSuccess(result: User?) {
            this.result = result
            finish()
        }

        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }

        private fun finish() {
            latch.countDown()
        }

        override fun onReadyToRender(myURLToRender: String) {
            this.myURLToRender = myURLToRender
            finish()
        }

    }

    @Test
    @Throws(IOException::class)
    fun testMICDefaultAPIVersion() {
        assertEquals("v3", client?.micApiVersion)
    }

    @Test
    @Throws(IOException::class)
    fun testMICCustomAPIVersion() {
        client?.micApiVersion = "1"
        assertEquals("v1", client?.micApiVersion)
    }

    companion object {
        private const val APP_KEY = "YOUR_APP_KEY_HERE"
        private const val APP_SECRET = "YOUR_APP_SECRET_HERE"
        private const val CLIENT_ID = "CLIENT_ID"
        private const val REFRESH_TOKEN = "refresh_token"
        private const val AUTH_TOKEN = "auth_token"
        private const val REDIRECT_URI = "http://test.redirect"
        private const val CLIENT_ID_FIELD = "client_id"
    }
}